doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+
+```cpp
+#define LAZY() if (cond)
+
+LAZY() expensiveFunctionCall();
+```
+
+In the example above `expensiveFunctionCall` is only executed when `cond` is `true`.
+If `cond` is a compile time constant set to `false`, the whole statement is compiled
+to a no-op and optimized away.
+
+This is the log macro with lazy evaluation
+
+```cpp
+#define IOX_LOG_INTERNAL(file, line, function, level)          \
+    if ((level) <= iox::log::Logger::minimalLogLevel()         \
+        && (iox::log::Logger::ignoreLogLevel()                 \
+            || (level) <= iox::log::Logger::getLogLevel()))    \
+    iox::log::LogStream(file, line, function, level).self()
+
+#define IOX_LOG(level) IOX_LOG_INTERNAL(__FILE__, __LINE__, __FUNCTION__, iox::log::LogLevel::level)
+```
+
+With `minimalLogLevel` and `ignoreActiveLogLevel` being static constexpr functions
+the compiler will optimize this either to `if (false) iox::log::LogStream(...)`
+and finally completely away or
+`if ((level) <= iox::log::Logger::activeLogLevel()) iox::log::LogStream(...)`.
+The minimal log level check is intended to fully optimize away a log statement
+and the ignore active log level check to always forward the log message to the
+logger, independent of the active log level.
+
+The `self()` invocation is used to create an lvalue reference to the `LogStream`
+object. This eases the implementation of logging support for custom types since
+`IOX_LOG(INFO) << myType;` would require to implement an overload with a rvalue
+`LogStream` reference but `IOX_LOG(INFO) << "#### " << myType;` requires a
+lvalue reference.
+
+The `IOX_LOG` macro is intended for general use and the `IOX_LOG_INTERNAL` for
+special cases when file, line and function are supplied from other places than
+the log macro invocation, like the `Expects` and `Ensures` macros.
+
+Although the macro contains an incomplete if-statement, the `LogStream` object at
+the end makes it safe to use since the compiler will complain if something else
+than a streaming operator or semicolon is used.
+
+#### Behavior before calling initLogger
+
+In order to have log messages before `initLogger` is called, the default logger
+is used with `LogLevel::INFO`. It is up to the implementation of the default
+logger what to do with these messages. For iceoryx the default logger is the
+`ConsoleLogger` (this can be changed via the platform abstraction) which will
+print the log messages to the console.
+
+#### Replacing the logger at runtime
+
+The default console logger can be replaced at runtime to a custom logger which
+can be used to forward the log messages to a logger of another framework.
+
+This is done by deriving from the base `Logger` class and implementing the pure
+virtual `setupNewLogMessage` and `flush` methods. Finally, the derived logger
+must be activated by calling the static `Logger::activeLogger` function and
+passing the new logger as argument to the function. The logger must have a static
```

Comment:
Yes, and it should be static.

Reply:
I agree. This makes management and runtime replacement of loggers easier. Then we can assume it exists for the whole lifetime of the program and do not have to worry about it. There are minor issues after main terminates (order of destruction), but those are manageable:

It is generally problematic if other objects with static lifetime use logging after main returns. The usage must ensure that the logger is destroyed last. However, we cannot enforce it is created first among all statics. It can be solved by constructing it into static memory (placement new) and never calling the dtor. Of course, the dtor does not have to be logically mandatory for this, e.g. have some connection to the outside world (e.g. close a data base connection).